lib/FetchEnvs.js

```diff
@@ -14,235 +14,195 @@
-        kubeError = e.message;
-      }
-      result = objectPath.get(res, ['data', key]);
-      type = objectPath.get(valueFrom, 'configMapKeyRef.type');
-    } else if (valueFrom['genericKeyRef']) {
-      let apiVersion = objectPath.get(valueFrom, 'genericKeyRef.apiVersion');
-      let kind = objectPath.get(valueFrom, 'genericKeyRef.kind');
-      let krm = await this.kubeClass.getKubeResourceMeta(apiVersion, kind, 'update');
-      let name = objectPath.get(valueFrom, 'genericKeyRef.name');
-      let namespace = objectPath.get(valueFrom, 'genericKeyRef.namespace', this.namespace);
-      let resource = {};
-      if (krm) {
-        try {
-          resource = await krm.get(name, namespace);
-        } catch (e) {
-          log.warn(e);
-          kubeError = e.message;
-        }
+        response = await this.api({
+          uri: krm.uri({ namespace, kind, name }),
+          json: true,
+          qs: matchLabelsQS
+        });
+      } catch (error) {
+        kubeError = error.message;
       }
-      key = objectPath.get(valueFrom, 'genericKeyRef.key');
-      result = objectPath.get(resource, ['data', key]);
-      type = objectPath.get(valueFrom, 'genericKeyRef.type');
     }
-    if (type && result) {
-      switch (type) {
-        case 'number':
-          result = Number(result);
-          break;
-        case 'boolean':
-          result = (result.toLowerCase() === 'true');
-          break;
-        case 'json':
-          result = JSON.parse(result);
-          break;
-        case 'jsonString':
-          // Stringify the jsonstring. This has the effect of double escaping the json, so that
-          // when we go to parse the final template to apply it to kube, it doesnt mistakenly
-          // turn our jsonString into actual json.
-          result = JSON.stringify(result);
-          // JSON.stringify adds quotes around the newly created json string. Kube forces us
-          // to wrap out curly braces in quotes so that it wont error on our templates. In order
-          // to avoid having 2 double quotes around the result, we need to remove the stringify
-          // quotes. slice(start of slice, end of slice)
-          result = result.slice(1, result.length - 1);
-          break;
-        case 'base64':
-          result = Buffer.from(result).toString('base64');
-          break;
+
+    let value = response?.data?.[key];
+
+    if (typeof matchLabelsQS === OBJECT) {
+      const output = response?.items.reduce(
+        reduceItemList(ref, strategy, decode),
+        Object.create(null)
+      );
+
+      value = output?.[key];
+      decode = false;
+    }
+
+    if (value === undefined) {
+      if (defaultValue === undefined) {
+        const msg = `failed to get env: ${JSON.stringify(conf)}. ${kubeError}`;
+        if (!optional) throw new Error(msg);
+        log.warn(msg);
+        this.updateRazeeLogs('warn', { controller: 'FetchEnvs', message: msg });
+      } else {
+        value = defaultValue;
```

Comment:
For secrets, my thought is if there is a default value, the user wouldnt define it as base64, so we need to make sure we dont decode the plain text defaultValue. do you agree?
```suggestion
        value = defaultValue;
        decode = false;
```

Reply:
agreed, but plain text or base64, is pretty much the same thing. so i think its a good idea to leave it plain text as a reminder that this isnt the place for secret data. maybe just some of default non secretive value to tell you down the pipe your secret is missing